File: android/app/src/main/kotlin/com/cruisemesh/app/relay/RelayPushBackoff.kt
package com.cruisemesh.app.relay

/**
 * Pure exponential-backoff decision logic for [RelayPushClient]'s reconnect
 * loop. Kept free of Android/OkHttp dependencies so it's unit-testable in
 * isolation (see RelayPushBackoffTest), mirroring
 * [com.cruisemesh.app.mesh.ReconnectBackoffTracker]'s BLE reconnect backoff --
 * but simpler: a relay endpoint doesn't go stale or rotate addresses the way
 * a BLE peer does, so there is no give-up threshold here. Every dropped
 * connection just doubles the wait (capped at [maxBackoffMs]); a successful
 * connect resets it back to [initialBackoffMs]. The 60s poll
 * (`RELAY_POLL_INTERVAL_MS` in `MeshService`) is the correctness backstop no
 * matter how long this backs off, so there is nothing worth giving up into.
 */
class RelayPushBackoff(
    private val initialBackoffMs: Long = INITIAL_BACKOFF_MS,
    private val maxBackoffMs: Long = MAX_BACKOFF_MS,
) {
    private var consecutiveFailures = 0

    /** Milliseconds to wait before the next reconnect attempt, given the current failure streak. */
    fun nextDelayMs(): Long {
        val shift = consecutiveFailures.coerceAtMost(20) // avoid Long overflow on shl
        return (initialBackoffMs shl shift).coerceAtMost(maxBackoffMs)
    }

    /** Record a failed/dropped connection, growing the next [nextDelayMs] result. */
    fun recordFailure() {
        consecutiveFailures++
    }

    /** Record a successful connect; resets the backoff back to [initialBackoffMs]. */
    fun recordSuccess() {
        consecutiveFailures = 0
    }

    companion object {
        const val INITIAL_BACKOFF_MS = 2_000L
        const val MAX_BACKOFF_MS = 60_000L
    }
}

File: ios/CruiseMesh/Relay/RelayPushBackoff.swift
import Foundation

/// Pure exponential-backoff decision logic for `RelayPushClient`'s reconnect
/// loop. Kept dependency-free so it's unit-testable in isolation (see
/// `RelayPushBackoffTests`), mirroring Android's `RelayPushBackoff`
/// (`android/.../relay/RelayPushBackoff.kt`) -- and, like it, deliberately
/// simpler than `ReconnectBackoffTracker`'s BLE reconnect backoff: a relay
/// endpoint doesn't go stale or rotate addresses the way a BLE peer does, so
/// there is no give-up threshold here. Every dropped connection just doubles
/// the wait (capped at `maxBackoffMs`); a successful connect resets it back
/// to `initialBackoffMs`. `MeshController`'s 60s poll (`runRelaySync`, driven
/// by `relayTimer`) is the correctness backstop no matter how long this backs
/// off, so there is nothing worth giving up into.
final class RelayPushBackoff {
    static let defaultInitialBackoffMs: Int64 = 2_000
    static let defaultMaxBackoffMs: Int64 = 60_000

    private let initialBackoffMs: Int64
    private let maxBackoffMs: Int64
    private var consecutiveFailures: Int = 0

    init(
        initialBackoffMs: Int64 = RelayPushBackoff.defaultInitialBackoffMs,
        maxBackoffMs: Int64 = RelayPushBackoff.defaultMaxBackoffMs
    ) {
        self.initialBackoffMs = initialBackoffMs
        self.maxBackoffMs = maxBackoffMs
    }

    /// Milliseconds to wait before the next reconnect attempt, given the current failure streak.
    func nextDelayMs() -> Int64 {
        let shift = min(consecutiveFailures, 20) // avoid overflow on the shift, mirrors the Kotlin port
        return min(initialBackoffMs << shift, maxBackoffMs)
    }

    /// Record a failed/dropped connection, growing the next `nextDelayMs` result.
    func recordFailure() {
        consecutiveFailures += 1
    }

    /// Record a successful connect; resets the backoff back to `initialBackoffMs`.
    func recordSuccess() {
        consecutiveFailures = 0
    }
}
